File: typescript/tests/storage-sqlite/trial-issuance.test.ts
/**
 * TrialIssuance CRUD on the SQLite adapter. Mirrors the memory-adapter
 * contract test; SQLite uses split partial unique indexes for the
 * NULLS-NOT-DISTINCT semantics so this exercise also pins the migration.
 */

import { Database } from 'bun:sqlite';
import { describe, expect, it } from 'bun:test';

import { SqliteStorage } from '../../src/storage/sqlite/index.ts';
import { applyMigrations } from '../../src/storage/sqlite/migrations.ts';
import type { UUIDv7 } from '../../src/types.ts';

const FP_A = 'a'.repeat(64);
const FP_B = 'b'.repeat(64);

function fresh(): { db: Database; s: SqliteStorage } {
  const db = new Database(':memory:');
  applyMigrations(db);
  const s = new SqliteStorage(db, { skipWalPragma: true });
  return { db, s };
}

async function freshTemplate(s: SqliteStorage, name: string): Promise<UUIDv7> {
  const tmpl = await s.createTemplate({
    scope_id: null,
    parent_id: null,
    name,
    max_usages: 5,
    trial_duration_sec: 86400,
    trial_cooldown_sec: 86400 * 7,
    grace_duration_sec: 0,
    force_online_after_sec: null,
    entitlements: {},
    meta: {},
  });
  return tmpl.id;
}

describe('sqlite adapter — TrialIssuance', () => {
  it('records and finds back', async () => {
    const { db, s } = fresh();
    const tmpl = await freshTemplate(s, 't1');
    const r = await s.recordTrialIssuance({ template_id: tmpl, fingerprint_hash: FP_A });
    const found = await s.findTrialIssuance({ template_id: tmpl, fingerprint_hash: FP_A });
    expect(found?.id).toBe(r.id);
    db.close();
  });

  it('rejects duplicate (template, fingerprint) via partial unique index', async () => {
    const { db, s } = fresh();
    const tmpl = await freshTemplate(s, 't1');
    await s.recordTrialIssuance({ template_id: tmpl, fingerprint_hash: FP_A });
    await expect(
      s.recordTrialIssuance({ template_id: tmpl, fingerprint_hash: FP_A }),
    ).rejects.toMatchObject({ code: 'UniqueConstraintViolation' });
    db.close();
  });

  it('NULL template_id is its own dedupe group', async () => {
    const { db, s } = fresh();
    await s.recordTrialIssuance({ template_id: null, fingerprint_hash: FP_A });
    await expect(
      s.recordTrialIssuance({ template_id: null, fingerprint_hash: FP_A }),
    ).rejects.toMatchObject({ code: 'UniqueConstraintViolation' });
    const ok = await s.recordTrialIssuance({ template_id: null, fingerprint_hash: FP_B });
    expect(ok.template_id).toBeNull();
    db.close();
  });

  it('deleteTrialIssuance frees the slot', async () => {
    const { db, s } = fresh();
    const tmpl = await freshTemplate(s, 't1');
    const first = await s.recordTrialIssuance({ template_id: tmpl, fingerprint_hash: FP_A });
    await s.deleteTrialIssuance(first.id);
    const second = await s.recordTrialIssuance({ template_id: tmpl, fingerprint_hash: FP_A });
    expect(second.id).not.toBe(first.id);
    db.close();
  });
});

File: licensing/storage/postgres/trial_issuance_test.go
package postgres_test

import (
	"errors"
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	"github.com/AnoRebel/licensing/licensing/storage/postgres"
)

const (
	fpA = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
	fpB = "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
)

func freshPgTemplate(t *testing.T, s *postgres.Storage, name string) string {
	t.Helper()
	tmpl, err := s.CreateTemplate(lic.LicenseTemplateInput{
		ScopeID:          nil,
		ParentID:         nil,
		Name:             name,
		MaxUsages:        5,
		TrialDurationSec: 86400,
		TrialCooldownSec: nil,
		GraceDurationSec: 0,
		Entitlements:     map[string]any{},
		Meta:             map[string]any{},
	})
	if err != nil {
		t.Fatal(err)
	}
	return tmpl.ID
}

func TestPostgres_TrialIssuance_RecordAndFind(t *testing.T) {
	s := freshPgStorage(t)
	tmpl := freshPgTemplate(t, s, "t1")
	r, err := s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: &tmpl, FingerprintHash: fpA})
	if err != nil {
		t.Fatal(err)
	}
	found, err := s.FindTrialIssuance(lic.TrialIssuanceLookup{TemplateID: &tmpl, FingerprintHash: fpA})
	if err != nil {
		t.Fatal(err)
	}
	if found == nil || found.ID != r.ID {
		t.Fatalf("find mismatch: %v vs %s", found, r.ID)
	}
}

func TestPostgres_TrialIssuance_RejectsDuplicate(t *testing.T) {
	s := freshPgStorage(t)
	tmpl := freshPgTemplate(t, s, "t1")
	_, err := s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: &tmpl, FingerprintHash: fpA})
	if err != nil {
		t.Fatal(err)
	}
	_, err = s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: &tmpl, FingerprintHash: fpA})
	var lerr *lic.Error
	if !errors.As(err, &lerr) || lerr.Code != lic.CodeUniqueConstraintViolation {
		t.Fatalf("expected UniqueConstraintViolation, got %v", err)
	}
}

func TestPostgres_TrialIssuance_NullTemplateGroup(t *testing.T) {
	s := freshPgStorage(t)
	_, err := s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: nil, FingerprintHash: fpA})
	if err != nil {
		t.Fatal(err)
	}
	_, err = s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: nil, FingerprintHash: fpA})
	if err == nil {
		t.Fatal("expected dup violation")
	}
	r, err := s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: nil, FingerprintHash: fpB})
	if err != nil || r.TemplateID != nil {
		t.Fatalf("unexpected: r=%v err=%v", r, err)
	}
}

func TestPostgres_TrialIssuance_DeleteFreesSlot(t *testing.T) {
	s := freshPgStorage(t)
	tmpl := freshPgTemplate(t, s, "t1")
	first, _ := s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: &tmpl, FingerprintHash: fpA})
	if err := s.DeleteTrialIssuance(first.ID); err != nil {
		t.Fatal(err)
	}
	second, err := s.RecordTrialIssuance(lic.TrialIssuanceInput{TemplateID: &tmpl, FingerprintHash: fpA})
	if err != nil {
		t.Fatal(err)
	}
	if second.ID == first.ID {
		t.Fatal("expected fresh id after delete + re-record")
	}
}
